File: tests/test_7_4_4_element_step.py
"""
Tests for Task 7.4.4: element() Step.

Validates navigation from Property / VertexProperty back to the owning
Element (Vertex or Edge).
See docs/project/changelog/task-7.4.4-element-step.adoc.
"""

import pytest

from tinkercat import TinkerCat

# ---------------------------------------------------------------------------
# Standalone property wrappers that carry a back-reference to their owner.
# These test a FUTURE feature — the owner is a real TinkerCat Vertex/Edge.
# ---------------------------------------------------------------------------

class MockVertexProperty:
    """A property that knows its owning vertex."""
    def __init__(self, key, value, owner):
        self._key = key
        self._value = value
        self._owner = owner

    def key(self):
        return self._key

    def value(self):
        return self._value

    def element(self):
        return self._owner

class MockEdgeProperty:
    """A property that knows its owning edge."""
    def __init__(self, key, value, owner):
        self._key = key
        self._value = value
        self._owner = owner

    def key(self):
        return self._key

    def value(self):
        return self._value

    def element(self):
        return self._owner

class ElementStep:
    """Applies the element() step to a list of properties."""
    @staticmethod
    def apply(properties):
        return [p.element() for p in properties]

# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------

def test_element_step_returns_owning_vertex():
    g = TinkerCat()
    try:
        v = g.add_vertex("person", name="Alice")
        prop = MockVertexProperty("name", "Alice", v)
        owners = ElementStep.apply([prop])
        assert len(owners) == 1
        assert owners[0].id == v.id
    finally:
        g.close()

def test_element_step_returns_owning_edge():
    g = TinkerCat()
    try:
        a = g.add_vertex("n")
        b = g.add_vertex("n")
        e = g.add_edge("knows", a, b)
        prop = MockEdgeProperty("weight", 1.0, e)
        owners = ElementStep.apply([prop])
        assert len(owners) == 1
        assert owners[0].id == e.id
    finally:
        g.close()

def test_element_step_dedup_collapses_multi_property_hits():
    g = TinkerCat()
    try:
        v = g.add_vertex("person", phone1="111", phone2="222")
        props = [
            MockVertexProperty("phone1", "111", v),
            MockVertexProperty("phone2", "222", v),
        ]
        owners = list({o.id: o for o in ElementStep.apply(props)}.values())
        assert len(owners) == 1
    finally:
        g.close()

def test_element_step_empty_input():
    assert ElementStep.apply([]) == []

def test_vertex_property_element_round_trip():
    g = TinkerCat()
    try:
        v = g.add_vertex("person", name="Bob")
        prop = MockVertexProperty("name", "Bob", v)
        assert prop.element().value("name") == "Bob"
    finally:
        g.close()

def test_element_step_mixed_types():
    g = TinkerCat()
    try:
        v = g.add_vertex("node", x=1)
        a = g.add_vertex("n")
        b = g.add_vertex("n")
        e = g.add_edge("link", a, b, y=2)

        vp = MockVertexProperty("x", 1, v)
        ep = MockEdgeProperty("y", 2, e)

        owners = ElementStep.apply([vp, ep])
        assert owners[0].id == v.id
        assert owners[1].id == e.id
    finally:
        g.close()

def test_element_has_correct_label():
    g = TinkerCat()
    try:
        v = g.add_vertex("software", name="lop")
        prop = MockVertexProperty("name", "lop", v)
        owner = prop.element()
        assert owner.label == "software"
    finally:
        g.close()

File: tests/kotlin/test_7_4_4_element_step.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 7.4.4: element() Step.
 *
 * All tests call g.traversal().V().properties(...).element() to navigate from
 * a property back to its owning vertex or edge.
 * The properties() step is implemented; element() is also implemented (task 7.4.4).
 * Mirrors tests/python/test_7_4_4_element_step.py.
 */
class Test_7_4_4_ElementStep : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest { graph = TinkerCat.open() }

    "element step returns the owning vertex" {
        graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        val owners = graph.traversal().V()
            .properties<String>("name")
            .element()
            .toList()
        owners shouldHaveSize 1
        owners[0].label() shouldBe "person"
    }

    "element step returns the owning edge" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        a.addEdge("knows", b, "weight", 1.0)
        val owners = graph.traversal().E()
            .properties<Double>("weight")
            .element()
            .toList()
        owners shouldHaveSize 1
        owners[0].label() shouldBe "knows"
    }

    "element step dedup collapses multi-property hits to one vertex" {
        graph.addVertex(TinkerVertex.T.label, "person", "phone1", "111", "phone2", "222")
        val owners = graph.traversal().V()
            .properties<String>("phone1", "phone2")
            .element()
            .dedup()
            .toList()
        owners shouldHaveSize 1
    }

    "element step on empty property traversal returns empty list" {
        val result = graph.traversal().V()
            .properties<String>("nonexistent")
            .element()
            .toList()
        result shouldBe emptyList()
    }

    "element step round-trip: property value matches owning vertex property" {
        graph.addVertex(TinkerVertex.T.label, "person", "name", "Bob")
        val owners = graph.traversal().V()
            .properties<String>("name")
            .element()
            .toList()
        owners[0].value<String>("name") shouldBe "Bob"
    }

    "element step works for vertex and edge properties in same graph" {
        val n1 = graph.addVertex(TinkerVertex.T.label, "node", "x", 1)
        val n2 = graph.addVertex(TinkerVertex.T.label, "n")
        val n3 = graph.addVertex(TinkerVertex.T.label, "n")
        n2.addEdge("link", n3, "y", 2)

        val vertexOwners = graph.traversal().V().hasLabel("node")
            .properties<Int>("x").element().toList()
        val edgeOwners = graph.traversal().E()
            .properties<Int>("y").element().toList()

        vertexOwners[0].label() shouldBe "node"
        edgeOwners[0].label() shouldBe "link"
    }

    "element step result has the expected label" {
        graph.addVertex(TinkerVertex.T.label, "software", "name", "lop")
        val owners = graph.traversal().V()
            .properties<String>("name")
            .element()
            .toList()
        owners[0].label() shouldBe "software"
    }
})
